lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,119 @@
+<?php
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_filter( 'block_core_navigation_render_inner_blocks', 'gutenberg_hook_blocks_into_core_navigation', 10, 3 );
+add_action( 'rest_insert_wp_navigation', 'gutenberg_update_ignore_hooked_blocks_meta_core_navigation', 10, 3 );
+
+/**
+ * Updates the post meta with the list of ignored hooked blocks when the navigation is created or updated via the REST API.
+ *
+ * @param WP_Post $post Post object.
+ */
+function gutenberg_update_ignore_hooked_blocks_meta_core_navigation( $post ) {
+	if ( ! isset( $post->ID ) ) {
+		return;
+	}
+
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $last_child_position ]
+	: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $first_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $first_child_position ]
+	: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, null );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, null );
+
+	$ignored_hooked_blocks = array_merge( $hooked_block_types_last_child, $hooked_block_types_first_child );
+
+	if ( ! empty( $ignored_hooked_blocks ) ) {
+		update_post_meta( $post->ID, '_wp_ignored_hooked_blocks', json_encode( $ignored_hooked_blocks ) );
+	}
+}
+
+/**
+ * Hooks into the REST API response for the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @param WP_REST_Request  $request  Request object.
+ * @return WP_REST_Response $response The response object.
+ */
+function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $post ) {
+	if ( ! isset( $response->data['content']['raw'] ) || ! isset( $response->data['content']['rendered'] ) ) {
+		return $response;
+	}
+	$parsed_blocks                         = parse_blocks( $response->data['content']['raw'] );
+	$content                               = serialize_blocks( gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post->ID ) );
```

Comment:
I've noticed that this serializes blocks that have previously been traversed-and-serialized, and the parsed, inside `gutenberg_hook_first_last_children_into_core_navigation` 😅 

Here's a diff to remove a few extraneous processing steps:

```diff
diff --git a/lib/compat/wordpress-6.5/navigation-block-hooks.php b/lib/compat/wordpress-6.5/navigation-block-hooks.php
index 038b1c31a37..7bd7e0d3c57 100644
--- a/lib/compat/wordpress-6.5/navigation-block-hooks.php
+++ b/lib/compat/wordpress-6.5/navigation-block-hooks.php
@@ -49,7 +49,13 @@ function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $p
 		return $response;
 	}
 	$parsed_blocks                         = parse_blocks( $response->data['content']['raw'] );
-	$content                               = serialize_blocks( gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post->ID ) );
+	$content                               = gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post->ID );
+
+	// Remove mock Navigation block wrapper.
+	$start   = strpos( $content, '-->' ) + strlen( '-->' );
+	$end     = strrpos( $content, '<!--' );
+	$content = substr( $content, $start, $end - $start );
+
 	$response->data['content']['raw']      = $content;
 	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
 
@@ -74,7 +80,10 @@ function gutenberg_hook_blocks_into_core_navigation( $inner_blocks, $post_id ) {
 		},
 		iterator_to_array( $inner_blocks )
 	);
-	return new WP_Block_List( gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post_id ) );
+
+	$markup        = gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post_id );
+	$parsed_blocks = parse_blocks( $markup );
+	return new WP_Block_List( $parsed_blocks[0]['innerBlocks'] );
 }
 
 /**
@@ -82,7 +91,7 @@ function gutenberg_hook_blocks_into_core_navigation( $inner_blocks, $post_id ) {
  *
  * @param array $inner_blocks Parsed inner blocks of the core/navigation block.
  * @param int $post_id Post ID of the navigation block.
- * @return array $inner_blocks Parsed inner blocks maybe with hooked blocks.
+ * @return string Parsed inner blocks maybe with hooked blocks.
  */
 function gutenberg_hook_first_last_children_into_core_navigation( $inner_blocks, $post_id = null ) {
 	$before_block_visitor = null;
@@ -111,9 +120,5 @@ function gutenberg_hook_first_last_children_into_core_navigation( $inner_blocks,
 		$before_block_visitor     = make_before_block_visitor( $hooked_blocks, null );
 		$after_block_visitor      = make_after_block_visitor( $hooked_blocks, null );
 	}
-	// TODO: We are unnecessarily serializing and deserializing the block here. Ideally we would just traverse without serializing.
-	$parsed_anchor_block_with_hooked_blocks = parse_blocks( traverse_and_serialize_block( $mock_anchor_parent_block, $before_block_visitor, $after_block_visitor ) );
-
-	// Note: If there are no hooked inner blocks, we return the original parsed blocks.
-	return isset( $parsed_anchor_block_with_hooked_blocks[0]['innerBlocks'] ) ? $parsed_anchor_block_with_hooked_blocks[0]['innerBlocks'] : $inner_blocks;
+	return traverse_and_serialize_block( $mock_anchor_parent_block, $before_block_visitor, $after_block_visitor );
 }
```

Reply:
Nice, thanks! I made a few further adjustments in `gutenberg_hook_first_last_children_into_core_navigation` as part of this to ensure we never end up with some undefined variables if the if condition is never met for whatever reason.

Updated in https://github.com/WordPress/gutenberg/pull/57754/commits/7546ba42970eb50a8c1380e15a17c8ac3452c477